### PR TITLE
feat(faiss): load a trained index and add a training endpoint

### PR DESCRIPTION
**Pull request by numb3r3**
One additional **training endpoint** is introduced in this PR. 

To use a trainable faiss indexer (e.g., IVF based):

**1)** we can first apply a training flow to train a faiss indexer (absolutely, user can also implement a locally training script using native faiss api):
```python
from jina import Flow
import numpy as np

train_filepath = 'train.npy'
train_data = np.array(np.random.random([10240, 256]), dtype=np.float32)
np.save(train_filepath, train_data)

f = Flow().add(
    uses="jinahub://FaissSearcher",
    timeout_ready=-1,
    uses_with={
        'index_key': 'IVF10_HNSW32,PQ64',
        'trained_index_file': 'faiss.index',
        'on_gpu': False,
    },
)

with f:
    # the trained index will be dumped to "faiss.index"
    f.post(on='/train', parameters={'train_filepath': train_filepath})
```

**2)** Then in the query runtime, we can use FaissSearch by providing a pre-trained index file resulted from step 1, e.g., 
```python
f = Flow().add(
    uses="jinahub://FaissSearcher",
    timeout_ready=-1,
    uses_with={
        'index_key': 'IVF10_HNSW32,PQ64',
        'trained_index_file': 'faiss.index', # trained indexer 
        'on_gpu': False,
        'dump_path': '/path/to/dump_file'
    },
)
```

One limitation to mention is:
> Concurrent search/add or add/add are not supported.
https://github.com/facebookresearch/faiss/issues/367

Hence the `index` operation can not utilize multi-core CPUs to speed up.

### CODE REVIEW

**Comment by JoanFM** [comment]
@numb3r3 concurrent operations should not be a problem, our parallelization model is based on multiprocess so one instance should have one index

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (JoanFM) [inline jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:59]: so we do not support anymore training at initialization right? it seems a good change to me

Comment (numb3r3) [reply jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:59]: yes, I created a separate training request endpoint to handle the training procedure.

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (maximilianwerk) [inline jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:179]: I think `Iterable` is a too weak requirement here, since you iterate over `vecs_iter` several times.

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (maximilianwerk) [inline jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:115]: That might be a stupid question due to my lack of understanding Faiss, but: Why do we need to build the index here? Shouldn't that be already part of the training? I would expect, that when I run the training method and restart the Searcher, it can directly pick up the trained AND filled index from disc. Or is this not possible/performant?

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (maximilianwerk) [inline jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:304]: Do I get the following right:

We would have a TrainingFlow running all the time (or when training is needed), which is able to react an `train` requests. It will train a Faiss index, fill it with the latest data and dump it somewhere (potentially to S3 or so). Then a running FaissSearcher can be restarted with the new dump path and after loading the index directly start serving data.

I like it. My first reaction was "That is a miss-use of the Jina Flow concept!! How dare you!", but actually it has quite some charm. :)

**Review by maximilianwerk — CHANGES_REQUESTED** [review]
Would it be possible to not only train, but also populate the index already during the `train` request?

Before merging, please wait for @cristianmtr feedback.

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (numb3r3) [reply jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:179]: Due to limited RAM, `Iterable` is chosen here to load large-scale data with a streaming method. To make the iteration safe, I iterate over the `vecs_iter` using batch mode (the batch_size is provided by the argument of `prefetch_size`)

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (numb3r3) [reply jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:115]: > Why do we need to build the index here? Shouldn't that be already part of the training? 

1) with providing the `trained_index_file`, we can load indexer from `trained_index_file`, and then start to build the indexer.
2) without providing the `trained_index_file`, there are two options:
    2.1  train the faiss indexer, and then use method 1) to build the indexer;
    2.2 start the query flow, the searcher will train the indexer by using the indexing data (loading from dump datafile). 

> I would expect, that when I run the training method and restart the Searcher, it can directly pick up the trained AND filled index from disc. 

Yes,  the current PR is actually implementing what you described above.

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (cristianmtr) [inline jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:22]: Remove this

**Comment by cristianmtr** [comment]
Can you update the README with some notes on this change?

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (cristianmtr) [inline jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:229]: Do we want to keep this?

**Review by cristianmtr — CHANGES_REQUESTED** [review]
Can you also add a full integration test that does

- start flow with empty index/ no training
- search -- no results
- train 
- search -- results!

And you can add it to the README as a showcase for how to use this feature.

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (numb3r3) [reply jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:229]: I think supporting training at initialization would be a good feature for some cases (e.g., small/media data). I would like to keep it.

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (maximilianwerk) [reply jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:188]: But you consume `vecs_iter` twice. Once in line 197 or 201 and once in the `_build_partial_index` call. You cannot consume an `Iterable` twice.

```def get_iter(count):
    for i in range(count):
        yield i


my_iter = get_iter(10)

first = list(my_iter)
second = list(my_iter)
print('first: ', first)
print('second: ', second)
```
outputs:
```
first:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
second:  []
```

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (maximilianwerk) [reply jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:124]: I am pretty sure it isn't. The `_build_index` is still called on every startup. So let me rephrase this question on a different level:

Imagine you have a trained faiss model without data. And you have the same model filled with data and dump the filled index to disc (model + data is dumped).

Are the following two operations equally fast:

A. Loading the faiss model and populate it with data from a dump
B. just load the model + filled index from disc. No need to populate it with data, since it is already populated.

Is the B even possible?

If both are equally fast, we might still want to do B to not stress the PSQL database when to many shards ramp up.

Furthermore, doing B + incremental update would allow us to do performant query while indexing with short index-query delay.

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (numb3r3) [reply jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:188]: The elements fetching from the first iteration in line 197 are temporally filled into the list `self._prefetch_data`. And before the second iteration happened in the `_build_partial_index`, I first consume the data from `self._prefetch_data`. So, we did not lose data.

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (numb3r3) [reply jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:124]: ah,  you are right. I got misunderstanding your question. Yes, doing B is a great feature, and is planned to approach it in the next PR.

File: jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py
Comment (maximilianwerk) [reply jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py:188]: Ah I see. That makes a whole lot of sense. Thanks for the explanation.

**Review by cristianmtr — APPROVED** [review]

**Pull request merged**
